Sources/NIOHPACK/HuffmanCoding.swift

```diff
@@ -0,0 +1,221 @@
+            if nbits < 8 {
+                state.buffer[state.offset] = UInt8(truncatingIfNeeded: code)
+                state.remainingBits = 8 - nbits
+                return
+            }
+            
+            // longer path for larger amounts
+            if nbits > 24 {
+                state.buffer[state.offset] = UInt8(truncatingIfNeeded: code >> 24)
+                nbits -= 8
+                state.offset += 1
+            }
+            
+            if nbits > 16 {
+                state.buffer[state.offset] = UInt8(truncatingIfNeeded: code >> 16)
+                nbits -= 8
+                state.offset += 1
+            }
+            
+            if nbits > 8 {
+                state.buffer[state.offset] = UInt8(truncatingIfNeeded: code >> 8)
+                nbits -= 8
+                state.offset += 1
+            }
+            
+            if nbits == 8 {
+                state.buffer[state.offset] = UInt8(truncatingIfNeeded: code)
+                state.offset += 1
+                state.remainingBits = state.offset == state.buffer.count ? 0 : 8
+            } else {
+                state.remainingBits = 8 - nbits
+                state.buffer[state.offset] = UInt8(truncatingIfNeeded: code)
+            }
+        }
+    }
+    
+    private mutating func ensureBitsAvailable(_ bits: Int, _ buffer: inout ByteBuffer) {
+        let bytesNeeded = bits / 8
+        if bytesNeeded <= buffer.writableBytes {
+            // just zero the requested number of bytes before we start OR-ing in our values
+            buffer.withUnsafeMutableWritableBytes { ptr in
+                ptr.baseAddress!.assumingMemoryBound(to: UInt8.self).assign(repeating: 0, count: bytesNeeded)
+            }
+            return
+        }
+        
+        // one extra byte to ensure we have space to fill at the end of the encoded string, if we need it
+        let neededToAdd = bytesNeeded - buffer.writableBytes
+        let newLength = buffer.capacity + neededToAdd
+        
+        // reallocate the buffer to ensure we have the room we need
+        buffer.changeCapacity(to: newLength)
+        
+        // now zero all writable bytes that we expect to use
+        buffer.withUnsafeMutableWritableBytes { ptr in
+            ptr.baseAddress!.assumingMemoryBound(to: UInt8.self).assign(repeating: 0, count: bytesNeeded)
+        }
+    }
+}
+
+/// Errors that may be encountered by the Huffman decoder.
+public enum HuffmanDecoderError : Error
+{
+    /// The decoder entered an invalid state. Usually this means invalid input.
+    case invalidState
+    /// The output data could not be validated as UTF-8.
+    case decodeFailed
+}
+
+/// A `HuffmanDecoder` is an idempotent class that performs string decoding using
+/// a static table of values. It maintains no internal state, so is thread-safe.
+public struct HuffmanDecoder {
+    /// The decoder table. This structure doesn't actually take up any space, I think?
```

Comment:
No need to leave the question in the code: you can just run the code quickly and have the initialiser of this structure print its size using `print(MemoryLayout<HuffmanDecoder>.size)`.